connectors/sources/redis.py

```diff
@@ -0,0 +1,283 @@
+        }
+
+    def _set_internal_logger(self):
+        self.redis_client.set_logger(self._logger)
+
+    async def _remote_validation(self):
+        """Validate configured databases
+        Raises:
+            ConfigurableFieldValueError: Unavailable services error.
+        """
+        invalid_type = []
+        invalid_db = []
+        msg = ""
+        if self.redis_client.database != ["*"]:
+            async with self.redis_client.client() as redis_client:
+                for db in self.redis_client.database:
+                    if not db.isdigit() or int(db) < 0:
+                        invalid_type.append(db)
+                    check_db = await self.redis_client.validate_database(
+                        db=db, redis_client=redis_client
+                    )
+                    if not check_db:
+                        invalid_db.append(db)
+                if invalid_type and invalid_db:
+                    msg = f"Database {', '.join(invalid_db)} are not available. Also database element should be integer. Please correct database name: {', '.join(invalid_type)}"
+                elif invalid_db:
+                    msg = f"Database {', '.join(invalid_db)} are not available."
+                elif invalid_type:
+                    msg = f"All database element should be integer. Please correct database name: {', '.join(invalid_type)}"
+                if msg:
+                    raise ConfigurableFieldValueError(msg)
+
+    async def validate_config(self):
+        """Validates whether user input is empty or not for configuration fields
+        Also validate, if user configured databases are available in Redis."""
+        await super().validate_config()
+        await self._remote_validation()
+
+    async def format_document(self, **kwargs):
+        """Prepare document for database records.
+
+        Returns:
+            document: Modified document.
+        """
+        doc_id = hash_id(f"{kwargs.get('db')}/{kwargs.get('key')}")
+        document = {
+            "_id": doc_id,
+            "key": kwargs.get("key"),
+            "value": str(kwargs.get("value")),
+            "size_in_bytes": kwargs.get("size"),
+            "database": kwargs.get("db"),
+            "key_type": kwargs.get("key_type"),
+            "_timestamp": iso_utc(),
+        }
+        return document
+
+    async def ping(self):
+        try:
+            async with self.redis_client.client() as redis_client:
+                await redis_client.ping()
+                self._logger.info("Successfully connected to Redis.")
+        except Exception:
+            self._logger.exception("Error while connecting to Redis.")
+            raise
+
+    async def get_key_metadata(self, redis_client, key, db):
+        key_type = await redis_client.type(key)
+        key_value = await self.redis_client.get_key_value(
+            redis_client=redis_client, key=key, key_type=key_type
+        )
+        key_size = await redis_client.memory_usage(key)
```

Comment:
This concerns me. This means for each key-value pair, we need 3 operations.

After a quick search, we can use `mget` to get values in batch, it seems there's no easy way to get the type and size in batch.

Reply:
yes, There is no method present which returns type and size in single method or batch.